src/tasks/SwiftPseudoterminal.ts

```diff
@@ -0,0 +1,91 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the VSCode Swift open source project
+//
+// Copyright (c) 2024 the VSCode Swift project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of VSCode Swift project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import * as vscode from "vscode";
+import { SwiftProcess } from "./SwiftProcess";
+
+export class SwiftPseudoterminal implements vscode.Pseudoterminal, vscode.Disposable {
+    private readonly writeEmitter: vscode.EventEmitter<string> = new vscode.EventEmitter<string>();
+    private readonly closeEmitter: vscode.EventEmitter<number | void> = new vscode.EventEmitter<
+        number | void
+    >();
+
+    constructor(
+        private swiftProcess: SwiftProcess,
+        private options: vscode.TaskPresentationOptions
+    ) {}
+
+    private disposables: vscode.Disposable[] = [];
+
+    open(initialDimensions: vscode.TerminalDimensions | undefined): void {
+        try {
+            // Convert the pty's events to the ones expected by the Tasks API
+            this.disposables.push(
+                this.swiftProcess.onDidSpawn(() => {
+                    // Display the actual command line that we're executing. `echo` defaults to true.
+                    if (this.options.echo !== false) {
+                        this.writeEmitter.fire(`> ${this.swiftProcess.commandLine}\n\n\r`);
+                    }
+                }),
+                this.swiftProcess.onDidWrite(data => {
+                    // The terminal expects a string that has "\n\r" line endings
+                    this.writeEmitter.fire(data.replace(/\n(\r)?/g, "\n\r"));
+                }),
+                this.swiftProcess.onDidThrowError(() => {
+                    this.closeEmitter.fire();
+                    this.dispose();
+                }),
+                this.swiftProcess.onDidClose(event => {
+                    this.closeEmitter.fire(event);
+                    this.dispose();
+                })
+            );
+            this.swiftProcess.spawn();
+            if (initialDimensions) {
+                this.setDimensions(initialDimensions);
+            }
+        } catch (error) {
+            this.closeEmitter.fire();
+            this.dispose();
+        }
+    }
+
+    dispose() {
+        for (const disposable of this.disposables) {
+            disposable.dispose();
+        }
+    }
+
+    handleInput(data: string): void {
+        const buf: Buffer = Buffer.from(data);
+        // Kill on ctrl+c
+        if (buf.length === 1 && buf[0] === 3) {
```

Comment:
Question, Is this called by vscode when you ctrl+c in the built in terminal? Is that the only way to receive this input sequence?

Reply:
Yep. VSCode calls handleInput, providing the buffer as a VT sequence. Can see https://vt100.net/docs/vt100-ug/chapter3.html. Look at Table 3-5  to see ctrl+c is 3